### DISCUSSION TITLE
Features for note taking in college?

### DISCUSSION THREAD

**Discussion by longtry** [xournalpp/xournalpp#4873] [Q&A]
Hi! I posted [a question](https://softwarerecs.stackexchange.com/questions/86998/an-open-source-note-taking-app-that-facilitates-drawing-handwriting) on stackexchange, but didn't receive any answer, probably because of the specificity of it. I did a bit of research myself and now think that _Xournal++_ is the brightest candidate, but I will need your help in confirming whether it can fulfill all the tasks in the link. (and if not, the solutions)

A bit of context: I'm not a tech geek at all, and new to (digital) note taking. I'm a Windows user but dipped into Linux very briefly in the past, so even as a newbie, I don't think I'll have much trouble with a new OS now. 1 of the features I want the app to have is spaced repetition reminder. That is, whenever I save a note after a lesson, by default it will be assigned reminders in a, b, c, d... days - according to my preprogrammed schedule. As long as the app is running in the background, it will put up alarms when the reminders are due, so I'll have to review the materials. 
If there are some important aspects that I should, um, take note of, please enlighten me.

Also, if you could go the extra mile and recommend the best hardware for the task, then I doubly appreciate that.

**Comment by rolandlo** [comment]
- [x] Pen/stylus support -> sure
- [ ] Process OCR -> not built in
- [x] Further programmed -> via plugins
- [ ] Easy to organize -> no, if you expect something like notebooks in OneNote. Organization is left to the file manager
- [x] Free -> definitely, 
- [ ] Flip horizontally -> you can easily flip strokes horizontally using a plugin, but not other stuff like background, images, text, TexImages, ...
- [x] customizable templates -> It's easy to change the paper size if you don't use a PDF background

About spaced repetition reminders: There is nothing built in like that. It would probably be easiest to do that externally, i.e. watch for file changes of .xopp-files and schedule a reminder

All in all, I'm not sure you'll be 100% happy with Xournal++ in regard of your special requirements. 
As for "best hardware": Not sure there is. Personally I'm happy with the Lenovo convertibles I have used, but other devices may work as well.

**Reply by longtry** [reply]
Thank you @rolandlo !
> Process OCR -> not built in

Come to think of it, OCR is actually not a "must", just "desired". Anyway, I read in [this link](https://alternativeto.net/software/xournal-/about/) that handwritten notes are converted to text, but I guess the info on that site is misleading. Does X++ have any OCR plugins?

> you can easily flip strokes horizontally using a plugin, but not other stuff like background, images, text, TexImages, ...

Can I save a note in .xopp and then flip the whole file? How many file types can X++ save?

> It's easy to change the paper size if you don't use a PDF background

Does this mean I can load jpg files like [this one](https://mthomearts.com/wp-content/uploads/2020/01/print-engineering-paper-c4fmjfrxadb5puzmajv4quzitggffupqyplgmjvfizc.jpg) and X++ will be able to recognize its components, i.e. the lines, and allow me to change them on the fly?

> About spaced repetition reminders: There is nothing built in like that. It would probably be easiest to do that externally

Doing that stuff manually will be a major pain, because my plan is to have 7-9 reminders for each lesson... I definitely need a plugin, then. Could you advise me on the difficulty level of building such an extension? If I make a request on the X++ community here, there's a chance someone will help... right? Sorry for being clueless, this is the 1st time I use github :)

I had a Lenovo laptop, pretty solid, but of the traditional type. Are the styluses in your convertible pen-sized? I think a diameter of at least 1cm is needed, otherwise it'll be too thin and hard to grasp for a prolonged time.

**Reply by rolandlo** [reply]
> Thank you @rolandlo !
> 
> > Process OCR -> not built in
> 
> Come to think of it, OCR is actually not a "must", just "desired". Anyway, I read in [this link](https://alternativeto.net/software/xournal-/about/) that handwritten notes are converted to text, but I guess the info on that site is misleading. Does X++ have any OCR plugins?
> 
The only thing that Xournal++ has is [this plugin](https://github.com/xournalpp/xournalpp/pull/2176) that uses a Google server to do the recognition. It's quite limited though.

> > you can easily flip strokes horizontally using a plugin, but not other stuff like background, images, text, TexImages, ...
> 
> Can I save a note in .xopp and then flip the whole file? How many file types can X++ save?

You could probably write a plugin doing that. Xournal++ saves in its own .xopp file format, but it can export to pdf, svg, png formats as well.
> 
> > It's easy to change the paper size if you don't use a PDF background
> 
> Does this mean I can load jpg files like [this one](https://mthomearts.com/wp-content/uploads/2020/01/print-engineering-paper-c4fmjfrxadb5puzmajv4quzitggffupqyplgmjvfizc.jpg) and X++ will be able to recognize its components, i.e. the lines, and allow me to change them on the fly?

No.
> 
> > About spaced repetition reminders: There is nothing built in like that. It would probably be easiest to do that externally
> 
> Doing that stuff manually will be a major pain, because my plan is to have 7-9 reminders for each lesson... I definitely need a plugin, then. Could you advise me on the difficulty level of building such an extension? If I make a request on the X++ community here, there's a chance someone will help... right? Sorry for being clueless, this is the 1st time I use github :)

I don't think you even need to write a plugin if you want to schedule reminders when saving the .xopp file. You can use `inotifywait` like [described here](https://github.com/xournalpp/xournalpp/issues/1028#issuecomment-825673215). The difficulty level depends a lot on what you expect from the reminders.
As for receiving help, yes, there's a change someone will help. It depends on how much your idea is interesting to other users though. So if it is designed very specific to your personal needs, then you probably won't receive much help.
> 
> I had a Lenovo laptop, pretty solid, but of the traditional type. Are the styluses in your convertible pen-sized? I think a diameter of at least 1cm is needed, otherwise it'll be too thin and hard to grasp for a prolonged time.

The stylus is pen-sized small and yes, a larger stylus feels better in the hand, but at least I don't loose it, since it fits very nicely into the chassis of the convertible. It's possible to use a larger stylus, it just won't be that easy to carry it along all the time.

**Reply by longtry** [reply]
> The only thing that Xournal++ has is https://github.com/xournalpp/xournalpp/pull/2176 that uses a Google server to do the recognition. It's quite limited though.

I just realized that you wrote that plugin, too. Cool!

> You could probably write a plugin doing that. Xournal++ saves in its own .xopp file format, but it can export to pdf, svg, png formats as well.

Since I know next to nothing about programming (my only few lessons back in the days were on... Pascal), I might ask someone. My impression is that, though X+ can export my fresh notes to .png, when I reopen that png, I'll only be able to add more content, right? Because my previous words are "fixed" pixels now and can't be extensively edited like .xopp.

> Does this mean I can load jpg files like [this one](https://mthomearts.com/wp-content/uploads/2020/01/print-engineering-paper-c4fmjfrxadb5puzmajv4quzitggffupqyplgmjvfizc.jpg) and X++ will be able to recognize its components, i.e. the lines, and allow me to change them on the fly?
No.

Then how would the task of changing paper size be carried out? Say I want to enlarge the paper. Will the letter I wrote be zoomed in, too? If they stay the same, how will their positions in regards to the paper be affected?

> I don't think you even need to write a plugin if you want to schedule reminders when saving the .xopp file. You can use inotifywait like https://github.com/xournalpp/xournalpp/issues/1028#issuecomment-825673215. The difficulty level depends a lot on what you expect from the reminders.

Here is the whole procedure I'm thinking of now for my case - it can be broadened for other people, too:
First, I set up in the tool/plugin option so that every time it runs, it create 6 reminders for a file, at 2, 6, 14, 30, 62, and 126 days after activation condition is met. The t/p is set to be activated whenever a new .xopp file is written on the disk (not when it's only edited). When a reminder is due, it could play sounds within the app, or even wake the device, depending on the setup. Extra feature if possible: the reminder can only be truly turned off when its target .xopp file is opened; otherwise we can only snooze it with a button.
Would the `inotifywait` tool be able to handle that? In case I need to ask for a plugin, would the above paragraph be adequate in describing the idea to people?

> As for receiving help, yes, there's a change someone will help. It depends on how much your idea is interesting to other users though.

Nice! So is it correct that what I'll need to do is press the "New pull request" button on the other page, and copy that paragraph there? I think that this function might be beneficial to any people who want to keep revisiting their ideas / want to organize their documents on a constant basis... not just college students. Out of curiosity, have you ever need to put a reminder to your notes?

> but at least I don't loose it, since it fits very nicely into the chassis of the convertible.

Good to know, thanks! BTW, what's the word for that elastic thingy that connects (secures) the pen to the device, or the stringy stuff that one puts around her wrist and tightens to keep the pen from falling? Losing my specific pen in a big uni environment will cause a headache for sure.

**Reply by longtry** [reply]
buzz

**Reply by rolandlo** [reply]
> Since I know next to nothing about programming (my only few lessons back in the days were on... Pascal), I might ask someone. My impression is that, though X+ can export my fresh notes to .png, when I reopen that png, I'll only be able to add more content, right? Because my previous words are "fixed" pixels now and can't be extensively edited like .xopp.
> 
That's correct.

 
> Then how would the task of changing paper size be carried out? Say I want to enlarge the paper. Will the letter I wrote be zoomed in, too? If they stay the same, how will their positions in regards to the paper be affected?
> 
No, the letters won't be scaled. They will be on the same lines as before, just some more lines will be added at the bottom. Just try it out (menu `Journal > paper format`)

> > I don't think you even need to write a plugin if you want to schedule reminders when saving the .xopp file. You can use inotifywait like [#1028 (comment)](https://github.com/xournalpp/xournalpp/issues/1028#issuecomment-825673215). The difficulty level depends a lot on what you expect from the reminders.
> 
> Here is the whole procedure I'm thinking of now for my case - it can be broadened for other people, too: First, I set up in the tool/plugin option so that every time it runs, it create 6 reminders for a file, at 2, 6, 14, 30, 62, and 126 days after activation condition is met. The t/p is set to be activated whenever a new .xopp file is written on the disk (not when it's only edited). When a reminder is due, it could play sounds within the app, or even wake the device, depending on the setup. Extra feature if possible: the reminder can only be truly turned off when its target .xopp file is opened; otherwise we can only snooze it with a button. Would the `inotifywait` tool be able to handle that? In case I need to ask for a plugin, would the above paragraph be adequate in describing the idea to people?
> 

It doesn't make sense to me to keep Xournal++ running while waiting for a reminder. I think your procedure should not be realized as a Xournal++ plugin, but as an external script that combines 

- `iwaitnotify` (for checking for newly created `.xopp` files)
- `crontab` (for scheduling a task)
- `notify-send` (for sending notifications)

Reading the manuals of these commands (just the part you need) would be a good start.

> > As for receiving help, yes, there's a change someone will help. It depends on how much your idea is interesting to other users though.
> 
> Nice! So is it correct that what I'll need to do is press the "New pull request" button on the other page, and copy that paragraph there? I think that this function might be beneficial to any people who want to keep revisiting their ideas / want to organize their documents on a constant basis... not just college students. Out of curiosity, have you ever need to put a reminder to your notes?
> 
No! "New pull request" is for actual code contributions to Xournal++, not for feature requests. I personally have never needed a reminder for my notes and you are the first one asking about that feature. I think keeping this discussion here is appropriate.
